### ISSUE TITLE
[no-use-before-declare] False positive for `keyof`/`typeof` parameter types

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [palantir/tslint#3655]
### Bug Report

- __TSLint version__: 5.9.1
- __TypeScript version__: 2.7.0-dev*
- __Running TSLint via__: (pick one) CLI / Node.js API / VSCode / grunt-tslint / Atom / Visual Studio / etc

#### TypeScript code being linted

```typescript
interface IGroupTypes {
    Scenery: { type: "scenery "},
    Solid: { type: "solid" },
}

function addToGroup(thing: IGroupTypes[typeof groupName], groupName: keyof IGroupTypes): void {
    // ...
}
```

with `tslint.json` configuration:

```json
{
    "rules": {
        "no-use-before-declare": true
    }
}
```

#### Actual behavior

```
ERROR: 6:47   no-use-before-declare  variable 'groupName' used before declaration
```

#### Expected behavior

There shouldn't be complaints about type parameters, as their types aren't runtime constructs.

**Comment by JoshuaKGoldberg** [comment]
Marking this as very low priority because `no-use-before-declare` is no longer a recommended rule. Sorry, @JoshuaKGoldberg from January!

**Comment by JoshuaKGoldberg** [comment]
This might be re-opened per discussion in #4789.
